net.rocketpool/image/src/app.ts

```diff
@@ -19,145 +23,118 @@ interface Event {
-          logDecoded.events.find((i) => i.name == "time").value
-        ),
-        transactionIndex: logRaw.transactionIndex,
-        transactionHash: logRaw.transactionHash,
-        logIndex: logRaw.logIndex,
-      };
-
-      if (!clb(event)) {
-        return;
-      }
-    }
-  } catch (err) {
-    // TODO: Is this too Infura-specific?
-    if (err.message.includes("query returned more than")) {
-      const middle = Math.round((fromBlock + toBlock) / 2);
-      console.log(
-        "Hit API limit, splitting interval to [%s, %s] and [%s, %s]",
-        fromBlock,
-        middle,
-        middle + 1,
-        toBlock
-      );
-
-      await readEventsFromETHRangeWithFallback(fromBlock, middle, clb);
-      await readEventsFromETHRangeWithFallback(middle + 1, toBlock, clb);
-    } else {
-      throw err;
-    }
+  if (isWebSocketSchema) {
+    return new ethers.providers.WebSocketProvider(providerUrl);
+  } else {
+    return new ethers.providers.JsonRpcProvider(providerUrl);
   }
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
 
+const env = cleanEnv(process.env, {
+  // Application:
+  ETH_NODE_PROVIDER_URL: url(),
+  // ODF interface:
+  ODF_BATCH_SIZE: num(),
+  ODF_NEW_ETAG_PATH: str(),
+  ODF_NEW_HAS_MORE_DATA_PATH: str(),
+})
+
 // See: https://etherscan.io/token/0xae78736cd615f374d3085123a210448e74fc6393
 const rethAddress = "0xae78736cd615f374d3085123a210448e74fc6393";
 const rethStartBlock = 13325304;
 
-// ABI
-const rethAbi = JSON.parse(fs.readFileSync("./abi/reth.abi.json", "utf8"));
-abiDecoder.addABI(rethAbi);
-
 // Init
-const providerUrl = process.env.ETH_NODE_PROVIDER_URL;
-console.log("ETH node provider URL: %s", providerUrl);
-const web3 = new Web3(
-  new Web3.providers.WebsocketProvider(providerUrl, {
-    clientConfig: {
-      maxReceivedFrameSize: 10000000,
-      maxReceivedMessageSize: 10000000,
-    },
-  })
-);
+const providerUrl = env.ETH_NODE_PROVIDER_URL;
+
+console.log(`ETH node provider URL: ${providerUrl}`);
+
+const provider = getProvider(providerUrl)
+const rethContract = RethAbi__factory.connect(rethAddress, provider)
 
 // Block range
-const etag = process.env.ODF_ETAG;
-const newEtagPath = process.env.ODF_NEW_ETAG_PATH;
+const lastSeenBlock = process.env.ODF_ETAG ? Number(process.env.ODF_ETAG): 0;
 
-const lastSeenBlock = etag.length != 0 ? Number(etag) : 0;
-console.log("Last seen block: %s", lastSeenBlock);
+console.log(`Last seen block: ${lastSeenBlock}`);
 
-const ethHeadBlock = await web3.eth.getBlockNumber();
-console.log("Current chain head block: %s", ethHeadBlock);
+const ethHeadBlock = await provider.getBlockNumber();
 
-let stride;
-if (process.env.STRIDE) {
-  stride = Number(process.env.STRIDE);
-} else {
-  stride = undefined;
-}
+console.log(`Current chain head block: ${ethHeadBlock}`);
+
+const startBlock = Math.max(lastSeenBlock + 1, rethStartBlock);
+const endBlock = Math.min(startBlock + env.ODF_BATCH_SIZE, ethHeadBlock);
```

Comment:
`ODF_BATCH_SIZE` is the recommended max number of records, but here you are using it for blocks.

I don't think we can establish any correlation between the number of blocks and number of events we expect to read from them.

I'd suggest to iterate block ranges with some fixed step (`stride` as it was done previously), and after each iteration checking if we have produced more than `ODF_BATCH_SIZE` events.

Reply:
A good catch -- reworked, done